### ISSUE TITLE
Variables used in computed method key are not scoped correctly

### ISSUE THREAD

**Issue by overlookmotel** [babel/babel#12790]
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

Where an object/class method has a computed key containing a variable, and the method includes a variable with same name in its body or parameters, the scope of the variable in method key is incorrectly identified.

Babel determines the scope of the var in the method key to be *inside* the method's body, not outside as it actually is.

**Input Code**

https://codesandbox.io/s/babel-issue-12790-b9y6b

Input code:

```js
const x = 'q';
const obj = {
  [x](x) {
    return x;
  }
};
```

Plugin:

```js
() => ({
  visitor: {
    Program(path) {
      path.scope.rename('x');
    }
  }
})
```

**Expected behavior**

Expected output to be:

```js
const _x = 'q';
const obj = {
  [_x](x) {
    return x;
  }
};
```

but instead get:

```js
const _x = 'q';
const obj = {
  [x](x) { // <-- `[x]` refers to non-existent variable
    return x;
  }
};
```

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**

None.

**Environment**

```
System:
  OS: macOS Mojave 10.14.6
Binaries:
  Node: 14.15.5 - ~/.nvm/versions/node/v14.15.5/bin/node
  npm: 6.14.11 - ~/.nvm/versions/node/v14.15.5/bin/npm
npmPackages:
  @babel/core: ^7.12.13 => 7.12.13
```

**Additional context**

I am very happy to work on a fix. But I'm unfamiliar with how/where Babel calculates scope of identifiers, so if someone could point me in the right direction, that would be a big help.

**Comment by char0n** [comment]
Yes I can confirm, I'm seeing this in multiple projects I work on. Babel no longer works on code like described in description of this issue.

Ref https://github.com/char0n/ramda-adjunct/pull/1750

**Comment by nicolo-ribaudo** [comment]
@char0n So a previous version of Babel used to work correctly?

**Comment by char0n** [comment]
@nicolo-ribaudo, yep @babel/preset-env@7.12.11 works like a charm. Anything with higher semver breaks.

**Comment by nicolo-ribaudo** [comment]
@char0n I think your bug is a different one, because the code in the OP also shows the bug with Babel 7.10.0.

**Comment by nicolo-ribaudo** [comment]
@overlookmotel It seems that there are two different bugs in Babel:
1. When `path` refers to the `x` inside the computed key, `path.scope` is the method scope rather than the program scope. This also causes scoping problems such as in [this example](https://babeljs.io/repl#?browsers=&build=&builtIns=false&spec=false&loose=true&code_lz=DYUwLgBAHg3AUHAxgewHYGdLIEYCsIC8EA3nBBANpQB0AngPzUBeAugBQCUJAvnN0A&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=Node-10&version=7.12.15&externalPlugins=)
2. When renaming a variable, we stop traversing the AST when we enter a node which introduces a new scope (such as `ObjectMethod`).

We can probably fix (1) by adding a list of "nodes whose keys should refer to the grandparent scope", and start from `path.parentPath` in https://github.com/babel/babel/blob/30dc25db0ea837181a4fdb2d82a376d5a4407c94/packages/babel-traverse/src/path/context.ts#L121 in those cases.

We can fix (2) by modifying https://github.com/babel/babel/blob/30dc25db0ea837181a4fdb2d82a376d5a4407c94/packages/babel-traverse/src/scope/lib/renamer.ts#L6 to do something similar to https://github.com/babel/babel/blob/30dc25db0ea837181a4fdb2d82a376d5a4407c94/packages/babel-helper-replace-supers/src/index.js#L47: when there is a computed method, we can use `skipAllButComputedKey` to continue visiting the method key.

**Comment by overlookmotel** [comment]
@nicolo-ribaudo Thanks for the diagnosis and hints. I'll make a stab at a fix in next couple of weeks.

**Comment by overlookmotel** [comment]
I've submitted PR #12812 to fix bug (1). It was more complicated than I expected, so I'm not sure now when I'll get time to attack bug (2).

**Comment by overlookmotel** [comment]
Fixed by #12812, now merged.

**Comment by nicolo-ribaudo** [comment]
Thanks @overlookmotel, we completely forgot to close this :sweat_smile:
